### ISSUE TITLE
branches_sharing_code reporting a false positive

### ISSUE THREAD

**Issue by MGlolenstine** [rust-lang/rust-clippy#7452]
I tried this code:

```rust
fn test() {
    let v = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
    let mut counter = 0;
    for i in v {
        if counter == 0 {
            counter += 1;
            println!("first");
        } else if counter == 1 {
            counter += 1;
            println!("second");
        } else {
            counter += 1;
            println!("other: {}", i);
        }
    }
}

```

I expected to see this happen: I expected the lint to pass without a hitch, as the counter is used in the `if` statement.

Instead, this happened: 
I got a lint report, that I could change the if statements, as they all start with the same line of code, but doing that makes the code invalid.
```
warning: all if blocks contain the same code at the start
 --> src/lib.rs:5:9
  |
5 | /         if counter == 0 {
6 | |             counter += 1;
  | |_________________________^
  |
  = note: `#[warn(clippy::branches_sharing_code)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#branches_sharing_code
help: consider moving the start statements out like this
  |
5 |         counter += 1;
6 |         if counter == 0 {
  |
```

### Meta

- `cargo clippy -V`: clippy 0.1.53 (53cb7b0 2021-06-17)
- `rustc -Vv`:
```
rustc 1.53.0 (53cb7b09b 2021-06-17)
binary: rustc
commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
commit-date: 2021-06-17
host: x86_64-unknown-linux-gnu
release: 1.53.0
LLVM version: 12.0.1
```

**Comment by MGlolenstine** [comment]
Suggesting label change to `I-suggestion-causes-error`, as it's correctly detected the case, but the fix causes a problem.

**Comment by xFrednet** [comment]
This kind of fits both labels, I'm going over a possible solution in my head and might claim this issue :)

**Comment by MGlolenstine** [comment]
@xFrednet Can this issue be closed now that the #7462 has been merged?
Thanks for the great work!

**Comment by xFrednet** [comment]
#7462 only documented this as a known issue. It's possible to actually fix this issue to not trigger the lint at all. I would keep this issue open until that is done, or it's decided that it's not worth it. 

You're welcome, thank you for the report. :upside_down_face:

**Comment by youknowone** [comment]
Similar false positive happens when the shared code is locking. Moving lock to outside of if statement may cause different behaviour.
